feat(ETH): eip1559 gas fee estimator and rpcs

dimxy opened the pull request:
Add eth EIP1559 priority fee estimator (WiP):

- [x] simple priority gas fee estimator based on eth_feeHistory call
- [x] loop running the estimator with blocktime period and storing values in a ctx
- [x] RPCs to start/stop loop for a enabled eth coin (ETH, USDT)
- [x] RPC to get estimated values

Add tx priority fee support for withdraw and swaps, improve gas limit for swap txns:

- [x] add eip1559 max priorty fee and max base fee v2 props to eth txns for withdraw and swaps
- [ ] consider adding GUI notifications based on websocket streaming channel (developed in PR #2041).
- [x] add alternative gas fee estimates from an API provider (a paid service) - added as test calls. In real it will be service on a mm2 eth provider
- [x] improve gas limit calculation to ask less gas from user: we cannot use eth_estimateGas for our etomic contract before swap started (as token spending approval is needed) so we use different consts for different swap operations
- [ ] do real-world test and adjust new gas limit consts if needed 
- [x] add tests for priority fee using the new dockerized geth node
- [x] merge eip1559 support in https://github.com/KomodoPlatform/rust-web3/pull/1 and https://github.com/KomodoPlatform/mm2-parity-ethereum/pull/1 and fix deps
- [x] add 'coins' param with transaction type (legacy, 1, 2) supported for different eth coins: ETH, ETC etc
- [x] decide where to have FeeEstimatorContext
- [ ] decide where to store gas api provider config (currently in MM2.json)
- [x] make config which chains are supported for gas estimation with provider and which use only internal estimator (added 'gas_fee_estimator' in coins)
- [ ] add eip1559 for trezor (when PR #1979 merged)
- [ ] improve gas fee estimates update: use eth_subscribe to get estimates when a new block comes, make a config with interval value if a gas api provides frequent updates 
- [ ] in this PR trade_preimage rpc now does not include refund fee. Make a option to return it or always return it in a dedicated field

This PR covers requirements in #1848 issue for Eth coin and ERC20 tokens, allowing more predictable priority-based gas fee and lesser gas limit for swap transactions

artemii235 reviewed (commented):
I have a question about how the feature is supposed to work in general: is it really required to run the background loop and keep additional context? What are the benefits of this approach over using "on demand" approach?

PS, by "on demand" I mean e.g., performing external API requests and all calculations in `get_eth_gas_price_estimated`.

cc @shamardy

dimxy commented:
> I have a question about how the feature is supposed to work in general: is it really required to run the background loop and keep additional context? What are the benefits of this approach over using "on demand" approach?

This is essentially how other wallets work. 
I believe calls to a gas api provider could last relatively long plus, if estimates are not expected to change until a new block comes it's no reason to ping it more often. Also the context with estimated data is shared among all eth coins/tokens.  Thus gui could fetch the recent results quickly, without forwarding calls to a provider

artemii235 commented:
I see, thanks for the explanation :slightly_smiling_face:

artemii235 commented on mm2src/coins/eth/eip1559_gas_fee.rs:
```diff
@@ -0,0 +1,469 @@
+    fn calculate_with_history(fee_history: &FeeHistoryResult) -> FeePerGasEstimated {
+        let base_fee = *fee_history.base_fee_per_gas.first().unwrap_or(&U256::from(0));
+        let base_fee = u256_to_big_decimal(base_fee, ETH_GWEI_DECIMALS).unwrap_or_else(|_| BigDecimal::from(0));
+        let mut priority_fees = vec![];
```
Using `vec![]` is overhead here, as it is always supposed to have `len == 3`. It can be a fixed-size array, pre-initialized with default values.

dimxy replied:
refactored this code

artemii235 reviewed (changes requested):
Next review iteration.

laruh commented:
@dimxy please keep in mind, that PR started to have conflicts in `eth.rs`

onur-ozkan commented on mm2src/coins/rpc_command/get_estimated_fees.rs:
```diff
@@ -0,0 +1,413 @@
+
+impl FeeEstimatorStartStopResponse {
+    #[allow(dead_code)]
+    pub fn get_result(&self) -> String { self.result.clone() }
```
nit suggestion: I typically prefer returning the reference of the value and letting the caller decide whether a clone is needed or not. For example, if caller needs only the reference of this value, with this approach you will clone it regardless.

dimxy replied:
changed to ref returned

onur-ozkan commented on mm2src/coins/rpc_command/get_estimated_fees.rs:
```diff
@@ -0,0 +1,413 @@
+#[derive(Debug, Display, Serialize, SerializeErrorType)]
+#[serde(tag = "error_type", content = "error_data")]
+pub enum FeeEstimatorError {
+    #[display(fmt = "Coin not activated or not a EVM coin")]
+    CoinNotFoundOrSupported,
```
I believe we should seperate these errors to be more as they are quite different scenarios (Not activated / Not an EVM coin)

dimxy replied:
done

onur-ozkan commented on mm2src/coins/rpc_command/get_estimated_fees.rs:
```diff
@@ -0,0 +1,413 @@
+        })))
+    }
+
+    /// Fee estimation update period in secs, basically equals to eth blocktime
+    fn get_refresh_interval() -> f64 { FEE_ESTIMATOR_REFRESH_INTERVAL as f64 }
```
Why not removing `FEE_ESTIMATOR_REFRESH_INTERVAL` and using const function?

like:

```
    const fn get_refresh_interval() -> f64 { 15. }
```

dimxy replied:
done

laruh commented on mm2src/coins/eth/eip1559_gas_fee.rs:
```diff
@@ -0,0 +1,469 @@
+use std::fmt;
+use web3::{types::BlockNumber, Transport};
+
+// Estimate base and priority fee per gas
+// or fetch estimations from a gas api provider
```
What is the purpose of this orphan comment?

dimxy replied:
made a source file doc comment instead

laruh commented on mm2src/coins/eth/eip1559_gas_fee.rs:
```diff
@@ -0,0 +1,469 @@
+
+const FEE_PER_GAS_LEVELS: usize = 3;
+
+#[derive(Clone, Debug, Serialize)]
+pub enum EstimationSource {
+    /// filled by default values
+    Empty,
+    Simple,
+    Infura,
+    Blocknative,
+}
```
Please provide doc com above `EstimationSource` and for each enumeration type

laruh commented on mm2src/coins/eth/eip1559_gas_fee.rs:
```diff
@@ -0,0 +1,469 @@
+    fn default() -> Self { Self::Empty }
+}
+
+#[derive(Clone, Debug, Serialize)]
+pub enum EstimationUnits {
```
please add doc comment for pub `EstimationUnits`

dimxy replied:
added

laruh reviewed (changes requested):
Thank you for this feature!
I have some notes for the first review iteration

dimxy commented on mm2src/coins/eth/eip1559_gas_fee.rs:
```diff
@@ -0,0 +1,469 @@
+            let max_fee_per_gas = base_fee.clone()
+                * BigDecimal::from_f64(Self::ADJUST_MAX_FEE[i]).unwrap_or_else(|| BigDecimal::from(0))
+                + max_priority_fee_per_gas.clone()
+                    * BigDecimal::from_f64(Self::ADJUST_MAX_PRIORITY_FEE[i]).unwrap_or_else(|| BigDecimal::from(0)); // TODO maybe use checked ops
```
this is to be tuned yet. Tests need to be done to adjust those coefficients for better estimations

laruh commented on mm2src/coins/eth/eip1559_gas_fee.rs:
Hmm I think it is worth splitting this file into three (or maybe more) modules, for eaxample: `infura_api`, `blocknative_api`, `gas_estimation`. So they will belong to `eip1559_gas_fee` module.

```
eth/
└── eip1559_gas_fee/
    ├── mod.rs          
    ├── infura_api.rs  
    ├── blocknative_api.rs
    └── gas_estimation.rs
```
    
Such structure will help us to keep code organized and maintainable, especially if code grows, or there will be changes in providers api. Each file focuses on a specific aspect of the functionality, making the codebase easier to navigate and understand.

dimxy replied:
I tried but realised I did not like it much: the code for gas api did not appear to me very significant (plus I had to copy many identical imports to both api files). 
Instead I created a mod for gas api calls in the same file.

laruh reviewed (changes requested):
some notes and suggestion related to module structure

dimxy commented on mm2src/coins/rpc_command/get_estimated_fees.rs:
```diff
@@ -0,0 +1,413 @@
+use std::pin::Pin;
+use std::sync::Arc;
+
+/// RPCs to start/stop gas price estimator and get estimated base and priority fee per gas
```
My intent was to add a comment to describe a purpose of the whole file. I guess I have seen similar comments in other sources but I can remove if this is against the rules

dimxy replied:
(maybe I should mark it with `//!`)

laruh replied:
> My intent was to add a comment to describe a purpose of the whole file. I guess I have seen similar comments in other sources but I can remove if this is against the rules

Oh, got your idea. 
In Rust, documentation for an entire module or file is typically written at the top of the file, inside a comment block that starts with `//!`. These are known as "outer" documentation comments and are used to document the module or crate as a whole. This contrasts with `///`, which are "inner" documentation comments used to document the item that immediately follows them, such as functions, structs, enums, constants, and other items within the code.

We have it in a lot of places in our repo, for example [here](https://github.com/KomodoPlatform/komodo-defi-framework/blob/976628b79c7b33301e03cb01f79ea4f4102a48df/mm2src/mm2_db/src/indexed_db/indexed_cursor.rs#L1)

laruh replied:
> (maybe I should mark it with `//!`)

yep yep :)

dimxy commented on mm2src/coins/rpc_command/get_estimated_fees.rs:
```diff
@@ -0,0 +1,413 @@
+    fn default() -> Self { Self::Stopped }
+}
+
+// Errors
```
Just wanted to separate parts of code in the source file. Will remove though

dimxy replied:
removed

dimxy commented on mm2src/coins/eth/eip1559_gas_fee.rs:
```diff
@@ -0,0 +1,469 @@
+
+#[allow(dead_code)]
+impl InfuraGasApiCaller {
+    const INFURA_GAS_API_URL: &'static str = "https://gas.api.infura.io/networks/1";
```
Another provider (which in fact we are going to use) supports only mainnet. I think won't support testnets for this feature. I added a check for chain_id in this commit [ee85d20](https://github.com/KomodoPlatform/komodo-defi-framework/pull/2051/commits/ee85d20d4d74fb6f10c600e1d4a4f74d1209b3f3)

dimxy commented on mm2src/coins/eth/eip1559_gas_fee.rs:
```diff
@@ -0,0 +1,469 @@
+
+    fn get_infura_gas_api_url() -> (String, Vec<(&'static str, &'static str)>) {
+        let url = format!("{}/{}", Self::INFURA_GAS_API_URL, Self::INFURA_GAS_FEES_CALL);
+        let headers = vec![("Authorization", INFURA_GAS_API_AUTH_TEST.as_str())];
```
added a config in MM2.json

dimxy commented on mm2src/coins/eth/eip1559_gas_fee.rs:
```diff
@@ -0,0 +1,469 @@
+        let resp = slurp_url_with_headers(&url, headers).await.mm_err(|e| e.to_string())?;
+        if resp.0 != StatusCode::OK {
+            let error = format!("{} failed with status code {}", Self::INFURA_GAS_FEES_CALL, resp.0);
+            info!("gas api error: {}", error);
```
This error is propagated to the GUI. Here this logging is for debug purposes (so I have just changed it to debug) - I think we do not want to flood logs with this kind of messages

dimxy commented on mm2src/coins/eth/eip1559_gas_fee.rs:
```diff
@@ -0,0 +1,469 @@
+            Self::BLOCKNATIVE_GAS_PRICES_CALL,
+            Self::BLOCKNATIVE_GAS_PRICES_PARAMS
+        );
+        let headers = vec![("Authorization", BLOCKNATIVE_GAS_API_AUTH_TEST.as_str())];
```
I added a config value for the gas api in MM2.json. However I'd like to leave the api key const like it is for now. This is a temp solution, for testing gas api, until we switch to use komodo proxy instead of direct connection to the api. With the komodo proxy the authentication would be different and this const would be removed

dimxy commented on mm2src/coins/eth/eip1559_gas_fee.rs:
```diff
@@ -0,0 +1,469 @@
+                Self::BLOCKNATIVE_GAS_PRICES_CALL,
+                resp.0
+            );
+            info!("gas api error: {}", error);
```
same as above, I suggest changing it to debug, not to flood log files. The error is sent to the gui nonetheless

dimxy commented on mm2src/coins/eth/eip1559_gas_fee.rs:
```diff
@@ -0,0 +1,469 @@
+            let max_priority_fee_per_gas = u256_to_big_decimal(max_priority_fee_per_gas, ETH_GWEI_DECIMALS)
+                .unwrap_or_else(|_| BigDecimal::from(0));
+            let max_fee_per_gas = base_fee.clone()
+                * BigDecimal::from_f64(Self::ADJUST_MAX_FEE[i]).unwrap_or_else(|| BigDecimal::from(0))
```
As I ve seen in another wallet impl similarly used values are different. So I did this the same way to tune the values after more tests are done

dimxy commented on mm2src/coins/eth/eip1559_gas_fee.rs:
```diff
@@ -0,0 +1,469 @@
+            base_fee,
+            low: priority_fees[0].clone(),
+            medium: priority_fees[1].clone(),
+            high: priority_fees[2].clone(),
```
refactored this code

dimxy commented on mm2src/coins/eth.rs:
```diff
@@ -4638,6 +4644,23 @@ impl EthCoin {
+        let history_estimator_fut = FeePerGasSimpleEstimator::estimate_fee_by_history(fee_history_namespace);
+        let provider_estimator_fut = BlocknativeGasApiCaller::fetch_blocknative_fee_estimation();
+        // To call infura use:
+        // let provider_estimator_fut = InfuraGasApiCaller::fetch_infura_fee_estimation();
```
added a config "gas_api" in MM2 json

dimxy commented on mm2src/coins/rpc_command/get_estimated_fees.rs:
```diff
@@ -0,0 +1,413 @@
+
+/// Get latest estimated fee per gas
+///
+/// Param: coin ticker
```
fixed

dimxy commented on mm2src/coins/rpc_command/get_estimated_fees.rs:
```diff
@@ -0,0 +1,413 @@
+    let coin = match coin {
+        MmCoinEnum::EthCoin(eth) => eth,
+        _ => return MmError::err(FeeEstimatorError::CoinNotFoundOrSupported),
+    };
```
fixed, thank for the improvement

dimxy commented on mm2src/coins/rpc_command/get_estimated_fees.rs:
```diff
@@ -0,0 +1,413 @@
+/// Stop gas priority fee estimator loop
+///
+/// Param: coin ticker
+pub async fn stop_eth_fee_estimator(ctx: MmArc, req: FeeEstimatorStartStopRequest) -> FeeEstimatorStartStopResult {
```
added

dimxy commented on mm2src/coins/eth/eip1559_gas_fee.rs:
```diff
@@ -0,0 +1,469 @@
+
+    /// estimate priority fees by fee history
+    fn calculate_with_history(fee_history: &FeeHistoryResult) -> FeePerGasEstimated {
+        let base_fee = *fee_history.base_fee_per_gas.first().unwrap_or(&U256::from(0));
```
I think it is very unlikely that the eth api returns incorrect data. But if it does the user in GUI would get zero base fee and should notice this or their txns would be failing.

(Also I have a TODO to add minimal max_fee_per_gas and max_priority_fee_per_gas (TBD yet) but I think they also should bet set to zero if incorrect data are fetched from an eth node)

artemii235 reviewed (changes requested):
I don't do it typically, but this time I thought that it is faster to write code rather than review comments :slightly_smiling_face: 

I have tried to simplify the `estimator_loop` as much as I could, please check the following diff (refactoring proposal): https://github.com/KomodoPlatform/komodo-defi-framework/compare/eip1559-gas-fee-estimate...KomodoPlatform:komodo-defi-framework:eip1559-gas-fee-estimate-refactoring-proposal?expand=1

dimxy commented:
> I have tried to simplify the `estimator_loop` as much as I could...

Thank you for this, studying...
I understand this code could be much simple. The use case I tried to implement was like that:
- Fee estimation loop starts when user opens a page with a transaction creation and finishes when this page is closed.
- Also, user could open concurrently 2 or more pages for different eth tokens (for tx creation), so the same estimation loop should be running and shared between those tokens. When all pages closed the loop ends.

If this is an overkill we can certainly reduce my code and just allow for GUI to handle estimation itself (when it needs it)

artemii235 commented:
@dimxy 
Yeah, I understand the idea. The proposal code actually implements it in a bit different way, considering the following:

1. Once the loop is started for ETH, it implicitly enables estimation for *all* ERC20 tokens, with `"platform_coin":"ETH"`. So there is no need to keep `using_coins` state in Rust. This should be documented, so GUI developers/users are aware. 
1. `run_state` can be simplified anyway because there seem to be only `Started` or `Stopped` state.
1. `abort_handler: AsyncMutex<Option<AbortOnDropHandle>>` covers both "stop by RPC" and "application stop" cases: it is explicitly taken from FeeEstimatorContext on RPC call or dropped with the context when `stop` RPC is invoked. I used `Arc<AsyncMutex<FeePerGasEstimated>>` as `fee_estimator_loop` argument to ensure that it doesn't keep the additional reference to FeeEstimatorContext, preventing it from dropping on `stop`.
1. Also, (just IMHO) wrapping `mpsc::Receiver` in `Arc<Mutex>` is a potential design problem. The `Single Consumer` pattern is not intended to be concurrently used in multiple threads (explicitly marked  `!Sync`).

dimxy commented:
> 4\. Also, (just IMHO) wrapping `mpsc::Receiver` in `Arc<Mutex>` is a potential design problem. The `Single Consumer` pattern is not intended to be concurrently used in multiple threads (explicitly marked  `!Sync`).

I understand now Arc is not good here. I guess I should have used simply Mutex. Thank you for that.

dimxy commented:
> @dimxy Yeah, I understand the idea. The proposal code actually implements it in a bit different way..

Yes I think it's better to use your code and let the GUI to control fee per gas estimation running

dimxy commented:
> > I understand now Arc is not good here. I guess I should have used simply Mutex. Thank you for that.
> 
> If possible, Mutex should be avoided too 🙂 The `Receiver` can be moved to the handling thread or future without wrappers. It has to be determined on case by case basis.

For future reference:
I believe this property !Sync for Receiver exists for std::sync::mpsc channel kind (and I used futures::channel::mpsc for which there is no such marker). 
I tried to make a test for std::sync::mpsc though: the compiler handles improper Receiver use pretty well, like wrapping it with Arc and sending Arc to a spawned thread. Besides, I guess !Sync does not prohibit sending Receiver itself to another thread (and compiler allowed to do this in my test). 
https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=82c1b216daddc1f104b15647739e4102)

shamardy commented on mm2src/coins/rpc_command/get_estimated_fees.rs:
```diff
@@ -0,0 +1,217 @@
+use std::sync::Arc;
+
+const FEE_ESTIMATOR_NAME: &str = "eth_gas_fee_estimator_loop";
+const ETH_SUPPORTED_CHAIN_ID: u64 = 1; // only eth mainnet is suppported (Blocknative gas platform currently supports Ethereum and Polygon/Matic mainnets.)
+                                       // To support fee estimations for other chains add a FeeEstimatorContext for a new chain
```
Maybe `FeeEstimatorContext` should be part of `EthCoin`/`EthCoinImpl` struct instead, and the `check_if_chain_id_supported` is done on coin initialization to initialize it for the supported chains. It would be good to not have the supported chains hardcoded too but to get if it's supported or not from coin config in case a new chain is supported.

dimxy replied:
I think we could create FeeEstimatorContext in the platform coin and set references to it when tokens are created.
As to have a setting if fee estimates are supported in the coins file would this not be error-prone to fix this setting for each token? Maybe better to have a whole gas provider api config, with chain id covering all tokens

shamardy replied:
> I think we could create FeeEstimatorContext in the platform coin and set references to it when tokens are created.

I agree.

> As to have a setting if fee estimates are supported in the coins file would this not be error-prone to fix this setting for each token? Maybe better to have a whole gas provider api config, with chain id covering all tokens

It should be for platform coin config only.

dimxy replied:
added to TODO in this PR

shamardy commented on mm2src/coins/rpc_command/get_estimated_fees.rs:
```diff
@@ -0,0 +1,217 @@
+    }
+
+    /// Fee estimation update period in secs, basically equals to eth blocktime
+    const fn get_refresh_interval() -> f64 { 15.0 }
```
If multiple coins are supported according to the previous comment, this should be block time, we have this in coin config called `avg_blocktime`. It might be removed anyway if we relied on `eth_subscribe` to get notifications for new headers/blocks.

dimxy replied:
yes I also was thinking about syncing fee updates with new block coming, thank you.
Though some gas fee providers provide more frequent and dynamic updates than each block, based on current mempool changes. For now I think we can stick with blocktime

shamardy replied:
> Though some gas fee providers provide more frequent and dynamic updates than each block, based on current mempool changes.

Please add a checklist in the original issue with this item and other items that we decide to do later or in next PRs. This will be a reminder to not close the issue until all things are resolved.

shamardy reviewed (commented):
Great work! This is a mid sprint review that focuses on providing early feedback that focus on functionalities more than code best practices. For fees for swaps, don't forget to implement eip1559 for trade_preimage and similar functions.

dimxy commented:
Latest commits:  
- added eip1559 priority fee to eth withdraw and swaps. Priority fee is set for eth coin or token with "set_swap_transaction_fee_policy" rpc
- added customised gas limit consts for eth swap operations
- changed tx fee in trade_preimage rpc (TradePreimageResponse::base_coin_fee and rel_coin_fee): now it does not include tx fee for possible refund. But taker_max_vol rpc still returns total tx fee to prevent too low balance to pay for refund (TODO: add a test for this)

Note: maybe in future we will want to have custom gas limit for different tokens if there are special recommendations. For now I found TUSD suggests gas limit of 150K and currently this value used for all erc20, so for this time we do not need this param yet

dimxy commented on mm2src/coins/eth.rs:
```diff
@@ -4741,6 +4752,39 @@ impl EthCoin {
+    /// Get gas base fee and suggest priority tip fees for the next block (see EIP-1559)
+    pub async fn get_eip1559_gas_fee(&self) -> Result<FeePerGasEstimated, MmError<Web3RpcError>> {
+        let coin = self.clone();
+        let fee_history_namespace: EthFeeHistoryNamespace<_> = coin.web3.api();
```
merged in [9347f10](https://github.com/KomodoPlatform/komodo-defi-framework/pull/2051/commits/9347f10fa6c5196b93ab3732a63614bf93df6e18)

laruh commented on mm2src/coins/Cargo.toml:
Why do you use links to your github acc and not to KomodoPlatform repo?

laruh reviewed (commented):
Thanks for the PR!
I have a question

dimxy commented:
> Thanks for the PR! I have a question

It's temporarily, I have added PRs with eip1559 support into our eth libs and temp use my repo until those PRs are merged

laruh commented on mm2src/coins/eth.rs:
```diff
@@ -6130,52 +6354,165 @@ impl From<Web3RpcError> for EthGasDetailsErr {
+    } else {
+        eth_value
+    };
+    let estimate_gas_req = CallRequest {
+        value: Some(eth_value_for_estimate),
+        data: Some(data),
+        from: Some(eth_coin.my_address),
+        to: Some(call_addr),
+        gas: None,
+        // gas price must be supplied because some smart contracts base their
+        // logic on gas price, e.g. TUSD: https://github.com/KomodoPlatform/atomicDEX-API/issues/643
+        gas_price: if let PayForGasOption::Legacy(LegacyGasPrice { gas_price }) = pay_for_gas_option {
+            Some(gas_price)
+        } else {
+            None
+        },
+        max_priority_fee_per_gas: if let PayForGasOption::Eip1559(Eip1559FeePerGas {
+            max_priority_fee_per_gas,
+            ..
+        }) = pay_for_gas_option
+        {
+            Some(max_priority_fee_per_gas)
+        } else {
+            None
+        },
+        max_fee_per_gas: if let PayForGasOption::Eip1559(Eip1559FeePerGas { max_fee_per_gas, .. }) = pay_for_gas_option
+        {
+            Some(max_fee_per_gas)
+        } else {
+            None
+        },
+        ..CallRequest::default()
+    };
```
please move fields creation before building CallRequest structure

```rust
    // gas price must be supplied because some smart contracts base their
    // logic on gas price, e.g. TUSD: https://github.com/KomodoPlatform/atomicDEX-API/issues/643
    let gas_price = match pay_for_gas_option {
        PayForGasOption::Legacy(LegacyGasPrice { gas_price }) => Some(gas_price),
        _ => None,
    };
    let max_priority_fee_per_gas = match pay_for_gas_option {
        PayForGasOption::Eip1559(Eip1559FeePerGas {
            max_priority_fee_per_gas,
            ..
        }) => Some(max_priority_fee_per_gas),
        _ => None,
    };
    let max_fee_per_gas = match pay_for_gas_option {
        PayForGasOption::Eip1559(Eip1559FeePerGas { max_fee_per_gas, .. }) => Some(max_fee_per_gas),
        _ => None,
    };
    let estimate_gas_req = CallRequest {
        value: Some(eth_value_for_estimate),
        data: Some(data),
        from: Some(eth_coin.my_address),
        to: Some(call_addr),
        gas: None,
        gas_price,
        max_priority_fee_per_gas,
        max_fee_per_gas,
        ..CallRequest::default()
    };
 ```
 Its hard to read when there is incorporation of if let expression (or match pattern) directly within the creation of a structure.

dimxy replied:
done in [8230a89](https://github.com/KomodoPlatform/komodo-defi-framework/pull/2051/commits/8230a8903b29f0f2492ccc4f101e4a72c09e09ab)
(yes this looks better, thanks)

laruh replied:
I would like to suggest to polish code a bit more. here I see that match pattern is more preferable

https://github.com/KomodoPlatform/komodo-defi-framework/blob/66f8ff76558e8330d08c6694d9e4efc6dbc3b7a3/mm2src/coins/eth.rs#L6434-L6448

```rust
    // Legacy transactions use a single gas price, while EIP-1559 transactions specify fees differently.
    let gas_price = match pay_for_gas_option {
        PayForGasOption::Legacy(LegacyGasPrice { gas_price }) => Some(gas_price),
        _ => None,
    };

    // Extract priority and max fees for EIP-1559 transactions, if applicable.
    let (max_priority_fee_per_gas, max_fee_per_gas) = match pay_for_gas_option {
        PayForGasOption::Eip1559(Eip1559FeePerGas {
            max_priority_fee_per_gas,
            max_fee_per_gas,
        }) => (Some(max_priority_fee_per_gas), Some(max_fee_per_gas)),
        _ => (None, None),
    };

    let estimate_gas_req = CallRequest {
        value: Some(eth_value_for_estimate),
        data: Some(data),
        from: Some(eth_coin.my_address),
        to: Some(call_addr),
        gas: None,
        // gas price must be supplied because some smart contracts base their
        // logic on gas price, e.g. TUSD: https://github.com/KomodoPlatform/atomicDEX-API/issues/643
        gas_price,
        max_priority_fee_per_gas,
        max_fee_per_gas,
        ..CallRequest::default()
    };
```

laruh commented on mm2src/coins/eth.rs:
```diff
@@ -2478,13 +2480,36 @@ async fn sign_and_send_transaction_with_metamask(
         to,
         gas: Some(gas),
-        gas_price: Some(gas_price),
+        gas_price: if let PayForGasOption::Legacy(LegacyGasPrice { gas_price }) = pay_for_gas_option {
+            Some(gas_price)
+        } else {
+            None
+        },
+        max_priority_fee_per_gas: if let PayForGasOption::Eip1559(Eip1559FeePerGas {
+            max_priority_fee_per_gas,
+            ..
+        }) = pay_for_gas_option
+        {
+            Some(max_priority_fee_per_gas)
+        } else {
+            None
+        },
+        max_fee_per_gas: if let PayForGasOption::Eip1559(Eip1559FeePerGas { max_fee_per_gas, .. }) = pay_for_gas_option
+        {
+            Some(max_fee_per_gas)
+        } else {
+            None
+        },
```
same, please move fields calculation before `let tx_to_send = TransactionReques` to improve readability

dimxy replied:
done

laruh replied:
here please use match pattern. 

PS: I there is a code duplication in three places. Please create priv function to create  gas price values and use it.

Here two. pllus one from this comment

https://github.com/KomodoPlatform/komodo-defi-framework/blob/66f8ff76558e8330d08c6694d9e4efc6dbc3b7a3/mm2src/coins/eth.rs#L768-L782

https://github.com/KomodoPlatform/komodo-defi-framework/blob/66f8ff76558e8330d08c6694d9e4efc6dbc3b7a3/mm2src/coins/eth.rs#L2506-L2520

```rust
            let gas_price = match pay_for_gas_option {
                PayForGasOption::Legacy(ref legacy_gas_price) => Some(legacy_gas_price.gas_price),
                _ => None,
            };

            let (max_priority_fee_per_gas, max_fee_per_gas) = match pay_for_gas_option {
                PayForGasOption::Eip1559(Eip1559FeePerGas {
                    max_priority_fee_per_gas,
                    max_fee_per_gas,
                }) => (Some(max_priority_fee_per_gas), Some(max_fee_per_gas)),
                _ => (None, None),
            };
```

dimxy replied:
fixed in [0ec27e9](https://github.com/KomodoPlatform/komodo-defi-framework/pull/2051/commits/0ec27e914bd8d3eb0dd40d1ff9500dad459ff742)

laruh commented on mm2src/coins/eth.rs:
```diff
@@ -2959,11 +2984,18 @@ impl EthCoin {
-                        // due to `u256_to_big_decimal` only.
-                        // Also TX history is not used by any GUI and has significant disadvantages.
-                        Some(EthTxFeeDetails::new(gas_used, gas_price, fee_coin).unwrap())
+                        let gas_price = web3_tx.gas_price.unwrap_or_default(); // TODO: create and use EthTxFeeDetails::from(web3_tx)
+                                                                               // It's relatively safe to unwrap `EthTxFeeDetails::new` as it may fail
+                                                                               // due to `u256_to_big_decimal` only.
+                                                                               // Also TX history is not used by any GUI and has significant disadvantages.
```
Please move todo comment and old comment `It's relatively safe to unwrap EthTxFeeDetails::new as it may fail...` above the code (dont inline it).

dimxy replied:
done [b21a186](https://github.com/KomodoPlatform/komodo-defi-framework/pull/2051/commits/b21a18681471c4f59d1382d22edabc7b0cc3d186)

laruh reviewed (changes requested):
some more notes

dimxy commented on mm2src/coins/rpc_command/get_estimated_fees.rs:
```diff
@@ -0,0 +1,226 @@
+use std::sync::Arc;
+
+const FEE_ESTIMATOR_NAME: &str = "eth_gas_fee_estimator_loop";
+const ETH_SUPPORTED_CHAIN_ID: u64 = 1; // only eth mainnet is suppported (Blocknative gas platform currently supports Ethereum and Polygon/Matic mainnets.)
```
done [8230a89](https://github.com/KomodoPlatform/komodo-defi-framework/pull/2051/commits/8230a8903b29f0f2492ccc4f101e4a72c09e09ab)

laruh replied:
yep, also write this long comment above the const

```rust
// Only eth mainnet is supported (Blocknative gas platform currently supports Ethereum and Polygon/Matic mainnets.)
// To support fee estimations for other chains add a FeeEstimatorContext for a new chain
const ETH_SUPPORTED_CHAIN_ID: u64 = 1
```

dimxy replied:
fixed [e473517](https://github.com/KomodoPlatform/komodo-defi-framework/pull/2051/commits/e473517812d3ca9c03ca45e7d340a4e48b0067be)

laruh commented:
:eyes: @dimxy 

Could you run `cargo clippy --all-targets -- -D warnings` please

I have errors locally
![image](https://github.com/KomodoPlatform/komodo-defi-framework/assets/52405288/dd9b1b63-470f-4e90-ac73-8c61c99a1bae)

laruh commented on mm2src/coins/eth.rs:
```diff
@@ -182,39 +199,27 @@ lazy_static! {
-    #[display(fmt = "Invalid response: {}", _0)]
-    InvalidResponse(String),
-    Internal(String),
+#[derive(Clone, Debug)]
+pub(crate) struct LegacyGasPrice {
+    pub gas_price: U256,
```
If `struct LegacyGasPrice` has `pub(crate)` visibility, its fields should have the same visibility level or more restrictive (like private). Please use `pub(crate) gas_price: U256,`

dimxy replied:
I thought maybe pub for members is okay here, like `pub(crate)` for the struct level defines visibility for the whole struct and `pub` for members simply to open/close access to specific member

laruh replied:
> I thought maybe pub for members is okay here, like `pub(crate)` for the struct level defines visibility for the whole struct and `pub` for members simply to open/close access to specific member

:D reminds me of some of my favorite threads in the CLI or electrum connections PR. I will try to find it, as there a lot of review notes.

my answer is this: its about consistency of visibility. the visibility of struct fields should generally be **as restrictive as**, or more restrictive than, the **struct's own** visibility to maintain consistency and encapsulation. While `pub` on fields within a `pub(crate) struct` doesnt expose them outside of the crate due to the struct's limited visibility, using `pub(crate)` for fields explicitly aligns with the struct's intended encapsulation boundary. Also You shouldnt trust compiler and allow fields to have higher level of visibility then structure.

laruh replied:
@dimxy Here we go https://github.com/KomodoPlatform/komodo-defi-framework/pull/1768#discussion_r1222889458

laruh commented on mm2src/coins/eth.rs:
```diff
@@ -182,39 +199,27 @@ lazy_static! {
 
-impl From<serde_json::Error> for GasStationReqErr {
-    fn from(e: serde_json::Error) -> Self { GasStationReqErr::InvalidResponse(e.to_string()) }
+#[derive(Clone, Debug)]
+pub(crate) struct Eip1559FeePerGas {
+    pub max_priority_fee_per_gas: U256,
+    pub max_fee_per_gas: U256,
```
same as above, if you use `pub(crate)` visibility level for struct, fields should have the same lvl or more restrictive.

dimxy replied:
done in [0ec27e9](https://github.com/KomodoPlatform/komodo-defi-framework/pull/2051/commits/0ec27e914bd8d3eb0dd40d1ff9500dad459ff742)

shamardy commented on mm2src/coins/Cargo.toml:
```diff
@@ -106,7 +106,8 @@ url = { version = "2.2.2", features = ["serde"] }
 # We don't need the default web3 features at all since we added our own web3 transport using shared HYPER instance.
-web3 = { git = "https://github.com/KomodoPlatform/rust-web3", tag = "v0.19.0", default-features = false }
+# web3 = { git = "https://github.com/KomodoPlatform/rust-web3", tag = "v0.19.0", default-features = false }
+web3 = { git = "https://github.com/dimxy/rust-web3", default-features = false, branch = "add-chain-id" }
```
You can use this release now https://github.com/KomodoPlatform/rust-web3/releases/tag/v0.20.0

dimxy replied:
fixed [342d3f4](https://github.com/KomodoPlatform/komodo-defi-framework/pull/2051/commits/342d3f4b8cfc50b9c4310ca262bd3314de0d2e9d)

shamardy commented on mm2src/coins/Cargo.toml:
```diff
@@ -44,9 +44,9 @@ derive_more = "0.99"
 ethabi = { version = "17.0.0" }
-ethcore-transaction = { git = "https://github.com/KomodoPlatform/mm2-parity-ethereum.git" }
+ethcore-transaction = { git = "https://github.com/dimxy/mm2-parity-ethereum.git", branch = "eip1559-support" }
 ethereum-types = { version = "0.13", default-features = false, features = ["std", "serialize"] }
-ethkey = { git = "https://github.com/KomodoPlatform/mm2-parity-ethereum.git" }
+ethkey = { git = "https://github.com/dimxy/mm2-parity-ethereum.git", branch = "eip1559-support" }
```
This PR https://github.com/KomodoPlatform/mm2-parity-ethereum/pull/1 is merged, you can use the original branch again.

dimxy replied:
fixed in [342d3f4](https://github.com/KomodoPlatform/komodo-defi-framework/pull/2051/commits/342d3f4b8cfc50b9c4310ca262bd3314de0d2e9d)

shamardy commented on mm2src/coins/eth.rs:
```diff
@@ -4251,22 +4309,25 @@ impl EthCoin {
-                ..CallRequest::default()
-            };
-            coin.estimate_gas_wrapper(estimate_gas_req)
-                .map_to_mm_fut(Web3RpcError::from)
-        }))
+        let fee_policy_for_estimate = get_swap_fee_policy_for_estimate(self.get_swap_transaction_fee_policy());
+        Box::new(
+            coin.get_swap_pay_for_gas_option(fee_policy_for_estimate)
+                .and_then(move |pay_for_gas_option| {
+                    let eth_value = U256::zero();
+                    let estimate_gas_req = CallRequest {
+                        value: Some(eth_value),
+                        data: Some(call_data),
+                        from: Some(coin.my_address),
+                        to: Some(contract_addr),
+                        ..CallRequest::default()
+                    };
+                    // gas price must be supplied because some smart contracts base their
+                    // logic on gas price, e.g. TUSD: https://github.com/KomodoPlatform/atomicDEX-API/issues/643
+                    let estimate_gas_req = call_request_with_pay_for_gas_option(estimate_gas_req, pay_for_gas_option);
+                    coin.estimate_gas_wrapper(estimate_gas_req)
+                        .map_to_mm_fut(Web3RpcError::from)
+                }),
+        )
     }
```
Please make this function async and remove `and_then` since it can be removed after making the function async.

shamardy commented on mm2src/coins/eth.rs:
```diff
@@ -4842,52 +4909,136 @@ impl EthCoin {
-                    debug!("Error {} on eth_feeHistory request", e);
-                    None
-                },
-            };
+            let eth_gas_price_fut = async {
+                match coin.gas_price().await {
+                    Ok(eth_gas) => Some(eth_gas),
+                    Err(e) => {
+                        error!("Error {} on eth_gasPrice request", e);
+                        None
+                    },
+                }
+            }
+            .boxed();
+
+            let eth_fee_history_price_fut = async {
+                match coin.eth_fee_history(U256::from(1u64), BlockNumber::Latest, &[]).await {
+                    Ok(res) => res
+                        .base_fee_per_gas
+                        .first()
+                        .map(|val| increase_by_percent_one_gwei(*val, BASE_BLOCK_FEE_DIFF_PCT)),
+                    Err(e) => {
+                        debug!("Error {} on eth_feeHistory request", e);
+                        None
+                    },
+                }
+            }
+            .boxed();
 
+            let (eth_gas_price, eth_fee_history_price) = join(eth_gas_price_fut, eth_fee_history_price_fut).await;
             // on editions < 2021 the compiler will resolve array.into_iter() as (&array).into_iter()
             // https://doc.rust-lang.org/edition-guide/rust-2021/IntoIterator-for-arrays.html#details
-            IntoIterator::into_iter([gas_station_price, eth_gas_price, eth_fee_history_price])
+            IntoIterator::into_iter([eth_gas_price, eth_fee_history_price])
                 .flatten()
                 .max()
                 .or_mm_err(|| Web3RpcError::Internal("All requests failed".into()))
         };
         Box::new(fut.boxed().compat())
     }
```
Again, this can be much simpler if made async.

dimxy replied:
done in [531f1af](https://github.com/KomodoPlatform/komodo-defi-framework/pull/2051/commits/531f1afb6832974583ae7110575769b952393e77)

shamardy commented on mm2src/coins/eth.rs:
```diff
@@ -6130,52 +6374,151 @@ impl From<Web3RpcError> for EthGasDetailsErr {
+        }) => {
+            let max_fee_per_gas = wei_from_big_decimal(&max_fee_per_gas, ETH_GWEI_DECIMALS)?;
+            let max_priority_fee_per_gas = wei_from_big_decimal(&max_priority_fee_per_gas, ETH_GWEI_DECIMALS)?;
+            if let EthGasLimitOption::Set(gas) = gas_limit {
+                return Ok((
+                    gas.into(),
+                    PayForGasOption::Eip1559(Eip1559FeePerGas {
+                        max_fee_per_gas,
+                        max_priority_fee_per_gas,
+                    }),
+                ));
+            } else {
+                // go to gas estimate code
+                PayForGasOption::Eip1559(Eip1559FeePerGas {
+                    max_fee_per_gas,
+                    max_priority_fee_per_gas,
+                })
```
Match should be used here instead of if else
```rust
            match gas_limit {
                EthGasLimitOption::Set(gas) => {
                    return Ok((
                        gas.into(),
                        PayForGasOption::Eip1559(Eip1559FeePerGas {
                            max_fee_per_gas,
                            max_priority_fee_per_gas,
                        }),
                    ));
                },
                EthGasLimitOption::Calc => {
                    // go to gas estimate code
                    PayForGasOption::Eip1559(Eip1559FeePerGas {
                        max_fee_per_gas,
                        max_priority_fee_per_gas,
                    })
                },
            }
```

dimxy replied:
done in [531f1af](https://github.com/KomodoPlatform/komodo-defi-framework/pull/2051/commits/531f1afb6832974583ae7110575769b952393e77)

shamardy commented on mm2src/coins/eth.rs:
```diff
@@ -451,15 +491,13 @@ pub struct EthCoinImpl {
     contract_supports_watchers: bool,
     web3_instances: AsyncMutex<Vec<Web3Instance>>,
     decimals: u8,
-    gas_station_url: Option<String>,
-    gas_station_decimals: u8,
-    gas_station_policy: GasStationPricePolicy,
```
So gas station will not be used anymore, I agree but hope that no body depends on this. c.c. @cipig

shamardy reviewed (commented):
First review iteration where I ask for some code cleaning :)

dimxy commented on mm2src/coins/eth.rs:
```diff
@@ -4842,52 +4909,136 @@ impl EthCoin {
     }
 
+    /// Get gas base fee and suggest priority tip fees for the next block (see EIP-1559)
+    pub fn get_eip1559_gas_fee(&self) -> Web3RpcFut<FeePerGasEstimated> {
```
done in [531f1af](https://github.com/KomodoPlatform/komodo-defi-framework/pull/2051/commits/531f1afb6832974583ae7110575769b952393e77)

dimxy commented on mm2src/coins/eth.rs:
```diff
@@ -5635,70 +5835,110 @@ pub fn wei_from_big_decimal(amount: &BigDecimal, decimals: u8) -> NumConversResu
     };
 
     Ok(try_s!(SignedEthTx::new(unverified)))
 }
```
applied in [531f1af](https://github.com/KomodoPlatform/komodo-defi-framework/pull/2051/commits/531f1afb6832974583ae7110575769b952393e77)

laruh commented on mm2src/coins/qrc20.rs:
```diff
@@ -1367,6 +1367,7 @@ impl MmCoin for Qrc20Coin {
+        include_refund_fee: bool,
     ) -> TradePreimageResult<TradeFee> {
         let decimals = self.utxo.decimals;
         // pass the dummy params
```
There is a code repetition related to dummy params in `get_sender_trade_fee` function and in `get_receiver_trade_fee`

![Screenshot 2024-04-09 at 17 25 37](https://github.com/KomodoPlatform/komodo-defi-framework/assets/52405288/3b99ea0c-cf9d-4906-9d85-3e9090c6e787)

I believe you can init them with standalone priv function, something like this: 
```rust
fn init_trade_fee_params() -> (u32, Vec<u8>, Vec<u8>, H160) {
    let timelock = now_sec_u32();
    let secret_hash = vec![0; 20];
    let swap_id = qrc20_swap_id(timelock, &secret_hash);
    let receiver_addr = H160::default();
    (timelock, secret_hash, swap_id, receiver_addr)
}
```

dimxy replied:
well, I guess this is not this PR code.
(btw those parts are not totally same)

laruh commented on mm2src/coins/qrc20.rs:
```diff
@@ -1398,14 +1399,18 @@ impl MmCoin for Qrc20Coin {
-                self.sender_refund_output(&self.swap_contract_address, swap_id, value, secret_hash, receiver_addr)?;
-            self.preimage_trade_fee_required_to_send_outputs(vec![sender_refund_output], &stage)
-                .await?
+        let total_fee = if include_refund_fee {
+            let sender_refund_fee = {
+                let sender_refund_output =
+                    self.sender_refund_output(&self.swap_contract_address, swap_id, value, secret_hash, receiver_addr)?;
+                self.preimage_trade_fee_required_to_send_outputs(vec![sender_refund_output], &stage)
+                    .await?
+            };
+            erc20_payment_fee + sender_refund_fee
+        } else {
+            erc20_payment_fee
```
I think it is more readable if you move `erc20_payment_fee + sender_refund_fee` calculation from if else

```rust
        // Optionally calculate refund fee.
        let sender_refund_fee = if include_refund_fee {
            let sender_refund_output =
                self.sender_refund_output(&self.swap_contract_address, swap_id, value, secret_hash, receiver_addr)?;
            self.preimage_trade_fee_required_to_send_outputs(vec![sender_refund_output], &stage)
                .await?
        } else {
            BigDecimal::from(0) // No refund fee if not included.
        };

        let total_fee = erc20_payment_fee + sender_refund_fee;
 ```

dimxy replied:
done in [789ec76](https://github.com/KomodoPlatform/komodo-defi-framework/pull/2051/commits/789ec7683331ef23b44441626440c220442d41d4)

laruh reviewed (changes requested):
some notes about code enhancements

dimxy commented on mm2src/coins/eth/eip1559_gas_fee.rs:
```diff
@@ -0,0 +1,499 @@
+            min_wait_time: None,
+            max_wait_time: None, // TODO: maybe fill with some default values (and mark them as uncertain)?
+        })
+    }
```
thanks, fixed in [789ec76](https://github.com/KomodoPlatform/komodo-defi-framework/pull/2051/commits/789ec7683331ef23b44441626440c220442d41d4)

dimxy commented on mm2src/coins/eth/eip1559_gas_fee.rs:
```diff
@@ -0,0 +1,499 @@
+            max_wait_time: None,
+        }
+    }
+}
```
done in [789ec76](https://github.com/KomodoPlatform/komodo-defi-framework/pull/2051/commits/789ec7683331ef23b44441626440c220442d41d4)

dimxy commented:
Above there are few fixes on review notes, also implemented couple of todos (coins params whether gas provider and which tx type are supported) plus some refactor of gas limit constants

laruh commented:
ps @dimxy pr started to have conflicts

shamardy commented:
@dimxy please fix conflicts in this PR so that we can continue with the review.

laruh reviewed (changes requested):
Thanks for the fixes! I I have one note for now that caught my eye.

dimxy commented on mm2src/coins/rpc_command/get_estimated_fees.rs:
```diff
@@ -0,0 +1,335 @@
+            },
+            FeeEstimatorConf::NotConfigured => Ok(Arc::new(FeeEstimatorState::CoinNotSupported)),
+        }
+    }
```
thanks,
done in [902c0e3](https://github.com/KomodoPlatform/komodo-defi-framework/pull/2051/commits/902c0e39cc90fcbb4a48a23fbdbbec07834fbf4e)

laruh commented:
@dimxy note that PR started to have conflicts 👀

dimxy commented:
> @dimxy note that PR started to have conflicts 👀

updated

shamardy commented on mm2src/coins/lp_coins.rs:
```diff
@@ -4753,6 +4846,12 @@ pub async fn my_tx_history(ctx: MmArc, req: Json) -> Result<Response<Vec<u8>>, S
     Ok(try_s!(Response::builder().body(body)))
 }
 
+/// `get_trade_fee` rpc implementation.
+/// There is some consideration about this rpc:  
+/// for eth coin this rpc returns max possible trade fee (estimated for maximum possible gas limit for any kind of swap).
+/// However for eth coin, as part of fixing this issue https://github.com/KomodoPlatform/komodo-defi-framework/issues/1848,
+/// `max_taker_vol' and `trade_preimage` rpc now return more accurate required gas calculations.
+/// So maybe it would be better to deprecate this `get_trade_fee` rpc
```
Add comment about this in documentation of changes for QA. I think `trade_preimage` and `max_taker_vol` are used currently instead of this but to make sure.

dimxy replied:
added (below https://github.com/KomodoPlatform/komodo-defi-framework/pull/2051#issuecomment-2118192179)

shamardy commented on mm2src/mm2_main/src/lp_swap/maker_swap.rs:
```diff
@@ -2235,7 +2237,7 @@ pub async fn maker_swap_trade_preimage(
     let preimage_value = TradePreimageValue::Exact(volume.to_decimal());
     let base_coin_fee = base_coin
-        .get_sender_trade_fee(preimage_value, FeeApproxStage::TradePreimage)
+        .get_sender_trade_fee(preimage_value, FeeApproxStage::TradePreimage, NO_REFUND_FEE)
```
Why refund fee is not included in maker trade preimage?

dimxy replied:
This is supposed to be one of the fixes for too high fees. I consulted with QA and they said that users may refuse from starting the swap when they see too big total fee for it.
So this change to create the trade preimage tx with no refund fee was to decrease the total fee shown to the user. I checked the GUI code and believe they use the trade preimage tx as a source to show swap fee to the user - so I think now user will fee only for the payment tx.
At the same time, the max_taker_vol rpc still includes refund fee, so the GUI can check it and will not allow to run a swap if user cannot have all needed funds for refund.

I guess this is debatable and easy to turn off/on in the code, I decided to add this for experimenting how it may work in the GUI.

shamardy replied:
> I guess this is debatable and easy to turn off/on in the code

We can turn it off/on using the trade preimage request instead. We can have this the default behaviour with the option to include refund fee if this parameter is true in the request. No need to fix this now but maybe in another PR after this is merged or when/if this feature was requested by QA/GUI.

dimxy replied:
yes, a param for the GUI to include or not refund fee would be better

dimxy replied:
I even think, it is better to add refund fee to swap data as a dedicated field (and also return it in the trade preimage response), because max swap volume is also checked by mm2 internally during the swap and better to include refund fee there

dimxy replied:
added an item into todo in the PR description

shamardy commented on mm2src/mm2_main/src/lp_swap/taker_swap.rs:
```diff
@@ -2477,7 +2479,7 @@ pub async fn taker_swap_trade_preimage(
     let preimage_value = TradePreimageValue::Exact(my_coin_volume.to_decimal());
     let my_coin_trade_fee = my_coin
-        .get_sender_trade_fee(preimage_value, stage)
+        .get_sender_trade_fee(preimage_value, stage, NO_REFUND_FEE)
```
Same here.

shamardy reviewed (approved):
Amazing Work 🔥 ! All below comments are non-blockers!
Please add comments documenting new coin configs and new APIs, or API changes for QA team. You can add them in the tracking issue or in this PR. You can follow the way I used here to document changes https://github.com/KomodoPlatform/komodo-defi-framework/pull/1962#issuecomment-2077894078, https://github.com/KomodoPlatform/komodo-defi-framework/pull/2014#issuecomment-2044027552

dimxy commented on mm2src/coins/rpc_command/get_estimated_fees.rs:
```diff
@@ -0,0 +1,336 @@
+        conf: &Json,
+        coin_type: &EthCoinType,
+    ) -> Result<Arc<FeeEstimatorState>, String> {
+        let fee_estimator_json = conf["gas_fee_estimator"].clone();
```
Actually we always use the simple estimator if provider fails. So if we change this estimator config we probably would need to change this logic too (like let GUI select 'use provider, then simple' or 'use only provider' or 'use simple'. Currently I have done logic like one popular wallet does it.
Maybe we can get response from the GUI or QA and refactor then

dimxy commented:
@KomodoPlatform/qa this PR adds support for eth [EIP-1559](https://eips.ethereum.org/EIPS/eip-1559) gas priority fee, for withdrawals and swaps.

**Withdrawals with new priority fee per gas**
For withdrawals a new WithdrawFee::EthGasEip1559 option is added with max_priority_fee_per_gas and max_fee_per_gas (in gwei). 
WithdrawFee::EthGasEip1559 also has 'gas_option' enum field with EthGasLimitOption::Set(u64) or EthGasLimitOption::Calc variants, to set custom gas limit or make the API calculate it. 
see [sample](https://github.com/KomodoPlatform/komodo-defi-framework/blob/dbe8de05b806f33bdefa4d5b396237a1fe60333c/mm2src/mm2_main/tests/mm2_tests/mm2_tests_inner.rs#L6638) 

**Setting swap policy to eip1559**
We can set to use eip1559 policy for swap transactions by a new 'set_swap_transaction_fee_policy' rpc with options of Internal, Low, Medium and High. 'Internal' means to create legacy transactions with gas price obtained from eth nodes (the default). 
There is also an rpc to query currently set policy 'get_swap_transaction_fee_policy'.

**Getting suggested fee per gas**
There is a gas fee estimator added to API, which gets suggested gas fees for low, medium, high levels either from a supported gas api providers (Infura or BlockNative) or may use mm2 internal "simple" gas fee estimator based on fee history.

To configure gas fee provider url a new param in MM2.json was added:
`"gas_api": {"provider": "Blocknative", "url": "https://api.provider.com"}` or `"gas_api": {"provider": "Infura", "url": "https://api.provider.com"}`
In coins file there is a new parameter added indicting which gas api provider is supported:
```
{
  "coin": "ETH",
  ...
  "gas_fee_estimator": estimator_type
}
```
where `estimator_type` could be `"provider"` or `"simple"`.
For "simple" estimator we do not need anything to add into MM2.json.

When gas fee estimator configured, GUI should start it with a "start_eth_fee_estimator" rpc with a "coin" param. (There is also "stop_eth_fee_estimator" to stop it when GUI does not need it). The estimator is always started for the platform coin (ETH) although in the start param it can be a token, so the platform coin must be enabled for estimator to work. The running estimator periodically queries the configured gas fee provider or calls the simple estimator, if the provider fails.

GUI can query the suggested gas fees with "get_eth_estimated_fee_per_gas" rpc (with "coin" param) which returns a serialised FeePerGasEstimated struct with low, medium and high max priority and total fees.

dimxy commented:
@KomodoPlatform/qa this PR changes trade fees as part of initiatives to solve problems with high transaction fees. 
This all affects ETH only.

**"trade_preimage" rpc does not include refund fee**
The `"trade_preimage"` rpc now does not include refund fee in the transaction preimage trade fee. This is intended not to show too big transaction fee to the user, what may discourage him from starting the swap.

At the same time `"max_taker_vol"` and `"max_maker_vol"` rpc still include refund fee (what will prevent user from creation of a swap for his whole amount which he won't be able to refund back).  

To discuss: maybe it is worth to add a param to the `"trade_preimage"` request to allow for GUI to choose to include or not the refund fee in the trade_fee. Or, maybe always return both trade and refund fees separately.

**"get_trade_fee" rpc unchanged**
Note:  there is yet another `"get_trade_fee"` rpc and it was not changed (calculated based on max trade gas of 150,000). Apparently this rpc is not used in the GUI.

**Gas limits adjusted for swap operations**
Instead of using a constant of 150,000 multiple constants were added for various swap stages and coins/tokens.
The new constants are [here](https://github.com/KomodoPlatform/komodo-defi-framework/blob/dbe8de05b806f33bdefa4d5b396237a1fe60333c/mm2src/coins/eth.rs#L226).

This all is experimental at this stage and needs to be tested and discussed.

dimxy commented on mm2src/coins/rpc_command/get_estimated_fees.rs:
```diff
@@ -0,0 +1,336 @@
+        })
+    }
+
+    type Error = MmError<NumConversError>;
```
done [20d16d9](https://github.com/KomodoPlatform/komodo-defi-framework/pull/2051/commits/20d16d99d0f92e3b138f6689721e515155563227)

dimxy commented on mm2src/coins/rpc_command/get_estimated_fees.rs:
```diff
@@ -0,0 +1,336 @@
+impl FeeEstimatorStartStopResponse {
+    #[allow(dead_code)]
+    pub fn get_result(&self) -> &str { &self.result }
+}
```
removed in [20d16d9](https://github.com/KomodoPlatform/komodo-defi-framework/pull/2051/commits/20d16d99d0f92e3b138f6689721e515155563227)

dimxy commented on mm2src/coins/eth/eip1559_gas_fee.rs:
```diff
@@ -0,0 +1,500 @@
+use web3::types::BlockNumber;
+
+pub(crate) use gas_api::BlocknativeGasApiCaller;
+#[allow(unused_imports)]
```
removed [20d16d9](https://github.com/KomodoPlatform/komodo-defi-framework/pull/2051/commits/20d16d99d0f92e3b138f6689721e515155563227)

dimxy commented on mm2src/coins/eth/eip1559_gas_fee.rs:
```diff
@@ -0,0 +1,500 @@
+        })
+    }
+
+    type Error = MmError<NumConversError>;
```
done [20d16d9](https://github.com/KomodoPlatform/komodo-defi-framework/pull/2051/commits/20d16d99d0f92e3b138f6689721e515155563227)

dimxy commented on mm2src/coins/eth/web3_transport/mod.rs:
```diff
@@ -128,6 +128,11 @@ pub struct FeeHistoryResult {
     pub oldest_block: U256,
     #[serde(rename = "baseFeePerGas")]
     pub base_fee_per_gas: Vec<U256>,
+
```
done [20d16d9](https://github.com/KomodoPlatform/komodo-defi-framework/pull/2051/commits/20d16d99d0f92e3b138f6689721e515155563227)

dimxy commented:
> Please add comments documenting new coin configs and new APIs, or API changes for QA team.

done in https://github.com/KomodoPlatform/komodo-defi-framework/pull/2051#issuecomment-2117979019

laruh reviewed (approved):
LGTM! Huge job!

cipig commented on mm2src/coins/eth.rs:
```diff
@@ -196,11 +222,39 @@ const GAS_PRICE_APPROXIMATION_PERCENT_ON_ORDER_ISSUE: u64 = 5;
+    pub const ETH_PAYMENT: u64 = 65_000;
+    /// Gas limit for swap payment tx with ERC20 tokens
+    /// real values are 98,9K
+    pub const ERC20_PAYMENT: u64 = 120_000;
```
we need to set this higher, see https://github.com/KomodoPlatform/komodo-defi-framework/issues/2135
maybe ERC20_RECEIVER_SPEND too

dimxy replied:
this definitely should be tuned.
Should we maybe add different gas limit sets for different coins?
Or, maybe allow to override gas limit in coins file
